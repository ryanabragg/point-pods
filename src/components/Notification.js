import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { withStyles } from '@material-ui/core/styles';

import Snackbar from '@material-ui/core/Snackbar';
import SnackbarContent from '@material-ui/core/SnackbarContent';
import IconButton from '@material-ui/core/IconButton';

import green from '@material-ui/core/colors/green';
import amber from '@material-ui/core/colors/amber';

import CheckCircleIcon from '@material-ui/icons/CheckCircle';
import WarningIcon from '@material-ui/icons/Warning';
import ErrorIcon from '@material-ui/icons/Error';
import InfoIcon from '@material-ui/icons/Info';
import CloseIcon from '@material-ui/icons/Close';

const NotificationContext = React.createContext();

function withNotification(Component) {
  return function NotificationEnabledComponent(props) {
    return (
      <NotificationContext.Consumer>
        {context => <Component {...props} notification={context} />}
      </NotificationContext.Consumer>
    );
  };
}

const styles = theme => ({
  margin: {
    margin: theme.spacing.unit,
  },
  success: {
    backgroundColor: green[600],
  },
  error: {
    backgroundColor: theme.palette.error.dark,
  },
  info: {
    backgroundColor: theme.palette.primary.dark,
  },
  warning: {
    backgroundColor: amber[700],
  },
  icon: {
    fontSize: 20,
  },
  iconVariant: {
    opacity: 0.9,
    marginRight: theme.spacing.unit,
  },
  message: {
    display: 'flex',
    alignItems: 'center',
  },
});

class Notification extends React.Component {
  state = {
    open: false,
    notification: {},
  };

  queue = [];

  icons = {
    success: CheckCircleIcon,
    warning: WarningIcon,
    error: ErrorIcon,
    info: InfoIcon,
  };

  handleClose = (event, reason) => {
    if(reason === 'clickaway')
      return;
    this.setState({ open: false });
    if(typeof this.state.notification.onClose === 'function')
      this.state.notification.onClose();
  };

  processQueue = () => {
    if (this.queue.length > 0) {
      this.setState({
        notification: this.queue.shift(),
        open: true,
      });
    }
  };

  handleExited = () => {
    this.processQueue();
  };

  addNotification = (message = '', variant = 'info', duration = 6000, onClose = null) => {
    return new Promise((resolve, reject) => {
      const notice = {
        key: new Date().getTime(),
        variant: [
          'success',
          'warning',
          'error',
          'info'
        ].includes(variant) ? variant : 'info',
        message: message,
        duration: duration || null,
        onClose: onClose,
      };
      this.queue.push(notice);
      if (this.state.open) {
        this.setState({ open: false });
      } else {
        this.processQueue();
      }
      resolve(notice);
    });
  }

  render() {
    const { classes, children, ...other } = this.props;
    const { key, variant, message, duration } = this.state.notification;
    const Icon = this.icons[variant || 'info'];

    return (
      <React.Fragment>
        <NotificationContext.Provider value={this.addNotification}>
          {children}
        </NotificationContext.Provider>
        <Snackbar
          key={key}
          anchorOrigin={{
            vertical: 'top',
            horizontal: 'center',
          }}
          open={this.state.open}
          autoHideDuration={duration || null}
          onClose={this.handleClose}
          onExited={this.handleExited}
        >
          <SnackbarContent
            className={classNames(classes[variant], classes.margin)}
            aria-describedby="notification-snackbar"
            message={
              <span id="notification-snackbar" className={classes.message}>
                <Icon className={classNames(classes.icon, classes.iconVariant)} />
                {message}
              </span>
            }
            action={[
              <IconButton
                key="close"
                aria-label="Close"
                color="inherit"
                className={classes.close}
                onClick={this.handleClose}
              >
                <CloseIcon className={classes.icon} />
              </IconButton>,
            ]}
            {...other}
          />
        </Snackbar>
      </React.Fragment>
    );
  }
}

Notification.propTypes = {
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(Notification);
export { NotificationContext, withNotification };
