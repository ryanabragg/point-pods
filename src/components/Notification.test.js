import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Notification, { NotificationContext, withNotification } from './Notification';

test('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Notification />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test('matches the prior snapshot', () => {
  const component = renderer.create(<Notification />);
  expect(component.toJSON()).toMatchSnapshot();
});

test('renders children in a context provider', () => {
  const component = shallow(<Notification>
      <div />
      <ul>
        <li>one</li>
        <li>two</li>
      </ul>
    </Notification>
  );
  expect(component.type().prototype.constructor.name).toBe('Notification');
  expect(component.dive().type()).toEqual(React.Fragment);
  expect(component.dive().childAt(0).type()['$$typeof']).toEqual(React.createContext().Provider['$$typeof']);
  expect(component.dive().childAt(0).childAt(0).type()).toBe('div');
  expect(component.dive().childAt(0).childAt(1).type()).toBe('ul');
});

test('provides a HOC to wrap components in the context consumer', () => {
  const Test = withNotification(props => (<i>Test</i>));
  const WithContext = props => (
    <NotificationContext.Provider value='test1'>
      <Test id='test2' />
    </NotificationContext.Provider>
  );
  const component = shallow(<Test />);
  expect(component.type()['$$typeof']).toEqual(React.createContext().Consumer['$$typeof']);
  const componentRenderer = renderer.create(<WithContext />);
  expect(componentRenderer.root.children[0].type.prototype.constructor.name).toBe('NotificationEnabledComponent');
  expect(componentRenderer.root.children[0].children[0].props.notification).toBe('test1');
  expect(componentRenderer.root.children[0].children[0].props.id).toBe('test2');
  expect(componentRenderer.root.children[0].children[0].children[0].type).toBe('i');
});

describe('actions', () => {
  test('add default notification');
  test('add success notification');
  test('add warning notification');
  test('add error notification');
  test('add info notification');
  test('notifications time out unless duration is null');
  test('onClose prop fires when notifications timeout or are closed');
});
