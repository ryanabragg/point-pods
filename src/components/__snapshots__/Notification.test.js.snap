// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the prior snapshot 1`] = `null`;
